allennlp/data/dataset_readers/sharded_dataset_reader.py

```diff
@@ -59,6 +65,16 @@ def __init__(self, base_reader: DatasetReader, **kwargs) -> None:
         # all of the instances within each shard are used.
         self.reader.manual_distributed_sharding = True
 
+        # ShardedDatasetReader is a wrapper for the original base_reader so some of the parameters like 'lazy'
+        # can be safely inherited. However, ShardedDatasetReader is a class instance of a DatasetReader as well.
+        # So we give priority to the parameters for the current instance stored in 'kwargs'.
+        # If not present, we check the ones in the base reader
+        for attr_name in _INHERITED_DATASET_PARAMS:
+            attr_val = getattr(self.reader, attr_name)
+            # copy over only shared attributes between the two classes
+            if attr_name in self.__dict__:
+                setattr(self, attr_name, kwargs.get(attr_name, attr_val))
```

Comment:
I think this is more complicated than it needs to be. I also think it would be more robust to set `lazy` on the call to `super().__init__(...)` on line 46 above, since the base class `DatasetReader` might have special logic in the `__init__` method that depends on the value of `lazy` (well, that's not the case right now, but it could be in the future).

So something like (right above line 46):

```python
if "lazy" not in kwargs:
    kwargs["lazy"] = base_reader.lazy
```